### PR TITLE
Chains and application links improvements

### PR DESCRIPTION
**Pull request by RiccardoM**
## Description
This PR fixes some bugs that I found while manually testing the Desmos <-> Band integration. 
It also implements some improvements that will be useful for third party tools like DJuno. 

## Checklist
- [ ] Targeted PR against correct branch.
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Wrote unit tests.
- [ ] Wrote integration tests (simulation & CLI). 
- [ ] Updated the documentation. 
- [ ] Added an entry to the `CHANGELOG.md` file.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.

### CODE REVIEW

File: proto/desmos/profiles/v1beta1/query.proto [hunk 1]
```diff
@@ -57,9 +57,32 @@ service Query {
         "/desmos/profiles/v1beta1/chain-links/{user}";
   }
 
+  // UserChainLinks queries chain links for the given user
+  rpc UserChainLink(QueryUserChainLinkRequest)
+      returns (QueryUserChainLinkResponse) {
```
Comment (dadamu) [inline proto/desmos/profiles/v1beta1/query.proto:62]:
```suggestion
  // UserChainLinks queries chain links for the given user
  rpc UserChainLinks(QueryUserChainLinksRequest)
      returns (QueryUserChainLinksResponse) {
```

**Review by dadamu — CHANGES_REQUESTED** [review]
Overall looks great to me. Just a couple of minor things.

File: proto/desmos/profiles/v1beta1/query.proto [hunk 2]
```diff
@@ -57,9 +57,32 @@ service Query {
...
+        "/desmos/profiles/v1beta1/chain-links/{user}/{chain_name}/{target}";
+  }
+
   // UserApplicationLinks queries application links for the given user
   rpc UserApplicationLinks(QueryUserApplicationLinksRequest)
       returns (QueryUserApplicationLinksResponse) {
     option (google.api.http).get = "/desmos/profiles/v1beta1/app-links/{user}";
   }
+
+  // UserApplicationLinks queries a single application link for a given user,
```
[the remark this answers is not in the document]
Comment (RiccardoM) [reply proto/desmos/profiles/v1beta1/query.proto:74]: I wouldn't change this, the documentation for this method continues on the line below (split by `make proto-format`) and I think it's more complete

**Review by dadamu — APPROVED** [review]
LGTM

**Comment by dadamu** [comment]
@RiccardoM Just found some error when testing because the codec of `KeeperTestSuite` in  `commen_test.go` is still `codec.BinaryMarshaler`. It should be change to `codec.Marshaler`, too.

File: x/profiles/keeper/msg_server_app_link.go [hunk 3]
```diff
@@ -19,6 +22,18 @@ func (k Keeper) LinkApplication(
 		return nil, err
 	}
 
+	// See if the link already exists
+	link, found, err := k.GetApplicationLink(ctx, msg.Sender, msg.LinkData.Application, msg.LinkData.Username)
+	if err != nil {
+		return nil, err
+	}
+
+	if found && link.IsVerificationOngoing() {
```
Comment (bragaz) [inline x/profiles/keeper/msg_server_app_link.go:31]: Shouldn't we also check if the link has already been verified?

**Review by bragaz — COMMENTED** [review]
Besides the doubt I exposed, everything else looks good!

File: x/profiles/keeper/msg_server_app_link.go (hunk 3 above)
Comment (RiccardoM) [reply x/profiles/keeper/msg_server_app_link.go:31]: The ability to start the verification all-over has been left open for two scenarios: 

1. The user has verified the link, and the link validity has later expired (in the future, app links will have a limited validity)
2. The verification process has ended with an error. 

In both cases, the link would appear as the verification has ended, but the user must be able to start from scratch with the same link.

Comment (bragaz) [reply x/profiles/keeper/msg_server_app_link.go:31]: @RiccardoM Cool! Thanks for the clarification.

**Review by bragaz — APPROVED** [review]

**Pull request merged**